source/host_resolver.c

```diff
@@ -88,50 +83,188 @@ int aws_host_resolver_record_connection_failure(struct aws_host_resolver *resolv
     return resolver->vtable->record_connection_failure(resolver, address);
 }
 
+/*
+ * Used by both the resolver for its lifetime state as well as individual host entries for theirs.
+ */
+enum default_resolver_state {
+    DRS_ACTIVE,
+    DRS_SHUTTING_DOWN,
+};
+
 struct default_host_resolver {
     struct aws_allocator *allocator;
-    struct aws_cache *host_table;
-    /* Note: This can't be an RWLock as even an LRU cache read is a modifying operation */
-    struct aws_mutex host_lock;
+
+    /*
+     * Mutually exclusion for the whole resolver, includes all member data and all host_entry_table operations.  Once
+     * an entry is retrieved, this lock MAY be dropped but certain logic may hold both the resolver and the entry lock.
+     * The two locks must be taken in that order.
+     */
+    struct aws_mutex resolver_lock;
+
+    /* host_entry->host_name (string) -> host_entry */
+    struct aws_hash_table host_entry_table;
+
+    enum default_resolver_state state;
+
+    /*
+     * Tracks the number of launched resolution threads that have not yet invoked their shutdown completion
+     * callback.
+     */
+    uint32_t pending_host_entry_shutdown_completion_callbacks;
 };
 
 struct host_entry {
+    /* immutable post-creation */
     struct aws_allocator *allocator;
     struct aws_host_resolver *resolver;
     struct aws_thread resolver_thread;
+    const struct aws_string *host_name;
+    int64_t resolve_frequency_ns;
+    struct aws_host_resolution_config resolution_config;
+
+    /* synchronized data and its lock */
     struct aws_mutex entry_lock;
+    struct aws_condition_variable entry_signal;
```

Comment:
debatable/suggestion (you will end up modifying like every line of the file):
```
struct synced {
    struct aws_mutex lock;
    struct aws_condition_variable entry_signal;
    struct aws_cache *aaaa_records;
    ...
    enum default_resolver_state state;
};
```

Reply:
Maybe someday